## 📑Lista 1 de Revisão da Prova de Python

#### 1. Uma plataforma de jogos registra o desempenho de um jogador ao longo de diferentes rodadas de uma partida.

#### Considere o programa a seguir:

In [5]:
rodadas = int(input())
saldo = int(input())
bonus = 0

for rodada in range(1, rodadas + 1):
    if rodada % 2 == 0:
        saldo = saldo + rodada * 2
        print(saldo)

    else:
        saldo = saldo - rodada
        print(saldo)

    if saldo >= 10:
        bonus = bonus + 1
        saldo = saldo - 3
        print(saldo)

    print(rodada, saldo, bonus)

print("Final", saldo, bonus)


7
1 7 0
11
8
2 8 1
5
3 5 1
13
10
4 10 2
Final 10 2


In [ ]:
n = 5
for i in range(1,n+1):
    

2
3
4
5
6


#### 2. Uma cooperativa utiliza um sistema simples para registrar solicitações de retirada de produtos de um estoque compartilhado.

#### Considere o programa a seguir:

In [2]:
estoque = int(input())
pedido = int(input())
atendidos = 0

while pedido != 0 and estoque > 0:
    if pedido <= estoque:
        estoque = estoque - pedido
        atendidos = atendidos + 1
        print("OK", estoque)
    else:
        estoque = estoque + 2
        print("NEGADO", estoque)
    pedido = int(input())

print("Fim", atendidos, estoque)


NEGADO 7
OK 3
OK 1
Fim 2 1


#### A-) 10; 4; 3; 0

- Saídas:
- 1º Loop: print("OK",6)
- 2º Loop: print("OK",3)
- 3º Loop: *pedido ==0 o while é encerrado!*
- Saída final:  print("FIM",2,3)

#### B-) 5;7;4;2;0

- Saídas:
- 1º Loop: print(Negado,7)
- 2º Loop: print(OK,3)
- 3º Loop: print(OK,1) --> *Pedido == 0 while se encerra antes de entrar no 4º Loop*
- Saída final:  print("FIM",2,1)

In [3]:
2//2

1

#### 3. Um sistema de identificação recebe uma palavra e gera registros intermediários durante o processamento do texto.

#### Considere o programa a seguir:

In [42]:
8//2

4

In [ ]:
texto = input()
novo = ""
contador = 0

for caractere in texto:
    if contador % 2 == 0:
        novo = caractere + novo
    else:
        novo = novo + caractere

    if caractere == "a":
        print("A", contador)

    contador = contador + 1
    print("Trecho", novo)

print("Texto", novo)
print("Qtd", contador)


A 0
Trecho a
Trecho ab
A 2
Trecho aab
Trecho aabc
Texto aabc
Qtd 4


#### 4. Um sistema de análise processa dados organizados em grupos e produz um indicador parcial para cada grupo analisado.

#### **Considere o programa a seguir:**

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_4_teste_mesa.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### **Respostas:**

### <u>Atenção ao Loop Interno! Como a função range do for vai de 1 até *(n -1)* É IMPORTANTE, que isso esteja estabelecido, como a questão pede loops aninhados, o for interno estabelecido é de (1, i+2) então por exemplo, na primeira rodada que  i =1, o for interno vai de 1, até 1+2=3, ou seja vai de 1 até 2 (3-1)</u> Isso é extremamente importante e me confundi!

#### Código:

O for i in range(1, n + 1) → i = 1, 2, 3
O for j in range(1, i + 2) → j varia conforme i.

-🔹 i = 1
j = 1 → valor = 1×1 = 1 (ímpar) → parcial = 0 − 1 = −1

j = 2 → valor = 1×2 = 2 (par) → parcial = −1 + 2 = 1
→ parcial = 1 < limite (5) → total = 0 − 1 = −1
Saída: (1, 1, −1)


- 🔹 i = 2
j = 1 → valor = 2×1 = 2 (par) → parcial = 0 + 2 = 2

j = 2 → valor = 2×2 = 4 (par) → parcial = 2 + 4 = 6

j = 3 → valor = 2×3 = 6 (par) → parcial = 6 + 6 = 12
→ parcial = 12 ≥ limite (5) → total = −1 + 12 = 11
Saída: (2, 12, 11)

- 🔹 i = 3
j = 1 → valor = 3×1 = 3 (ímpar) → parcial = 0 − 1 = −1

j = 2 → valor = 3×2 = 6 (par) → parcial = −1 + 6 = 5

j = 3 → valor = 3×3 = 9 (ímpar) → parcial = 5 − 3 = 2

j = 4 → valor = 3×4 = 12 (par) → parcial = 2 + 12 = 14
→ parcial = 14 ≥ limite (5) → total = 11 + 14 = 25
Saída: (3, 14, 25)

#### Print final: Final 25

#### Letra b-):
#### -n=4 e limite =20

#### i = 1

- O loop interno roda com j = 1 e j = 2.

- O parcial termina em 1.

- Como 1 < 20, o total passa de 0 para −1.

- Saída: (1, 1, -1)

#### i = 2

- O loop interno roda com j = 1, 2, 3.

- O parcial termina em 12.

- Como 12 < 20, o total passa de −1 para −2.

#### Saída: (2, 12, -2)

#### i = 3

- O loop interno roda com j = 1, 2, 3, 4.

- O parcial termina em 14.

- Como 14 < 20, o total passa de −2 para −3.

#### Saída: (3, 14, -3)

#### i = 4

- O loop interno roda com j = 1, 2, 3, 4, 5.

- O parcial termina em 60.

#### Como 60 ≥ 20, o total passa de −3 para 57.

#### Saída: (4, 60, 57)

### Saída Final: (Final, 57)

#### 5. Um equipamento realiza várias etapas de verificação antes de concluir uma operação.

#### Em cada etapa, um valor numérico é recebido pelo sistema.

#### Considere o programa a seguir:

<div align="center">
<img 
    src="print_abertas_lista_1_02/quetao_5.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

In [11]:
etapas = int(input())
saldo = 0

for etapa in range(1, etapas + 1):
    valor = int(input())
    repeticoes = 0

    while valor > 0 and repeticoes < 3:
        if valor % 2 == 0:
            saldo = saldo + valor % 5
        else:
            saldo = saldo - 1

        valor = valor // 2
        repeticoes = repeticoes + 1

    if saldo < 0:
        saldo = etapa

    print("Etapa", etapa, saldo, repeticoes)

print("Saldo", saldo)


Etapa 1 9 3
Etapa 2 6 3
Etapa 3 7 2
Saldo 7


#### **6.** Um módulo de auditoria recebe uma sequência de símbolos e dois valores de configuração antes de produzir uma assinatura de processamento.

#### Considere o programa a seguir:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_6_teste_mesa.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

In [2]:
1%2

1

In [3]:
 0 > 0

False

In [2]:
codigo = input()
limite = int(input())
saldo = int(input())

resultado = ""
pos = 0

for caractere in codigo:
    repeticao = pos + 1

    while repeticao > 0:
        if (saldo + repeticao) % 2 == 0:
            resultado = resultado + caractere
            saldo = saldo + 1
        else:
            resultado = caractere + resultado
            saldo = saldo - 1

        repeticao = repeticao - 2

    if saldo > limite:
        print("P", pos, saldo)
        saldo = saldo - limite
    else:
        print("N", pos, saldo)

    pos = pos + 1

print("Resultado", resultado)
print("Saldo", saldo)


N 0 -1
N 1 -2
N 2 -2
N 3 -2
Resultado DCBACD
Saldo -2


#### 7. O calendário gregoriano utiliza uma regra especial para determinar os anos que possuem 366 dias. Um ano é bissexto quando é divisível por 4, exceto os anos divisíveis por 100, que somente são bissextos quando também são divisíveis por 400.

#### Faça um programa que leia um ano inteiro positivo e informe se ele é bissexto ou não bissexto.

#### Exemplos:
- • 2024 é bissexto;
- • 1900 não é bissexto;
- • 2000 é bissexto;
- • 2025 não é bissexto.

In [10]:
1900%4

0

In [23]:
def ano_bissexto(ano):
    if ano%4 ==0:
        if  ano%100 ==0:
            if ano% 400 ==0:
                return f"{ano} é bissexto!"
            else:
                return f"{ano} não é bissexto!"
        else:
            return f"{ano} é bissexto!"
    else:
        return f"{ano} não é bissexto!"

In [28]:
ano = int(input("Digite um ano para saber se é bissexto ou não: "))
while ano <0:
    print(f"A entrada {ano} deve ser maior que zero")
    ano = int(input("Digite um ano para saber se é bissexto ou não: "))
print(ano_bissexto(ano))
    

1900 não é bissexto!


#### **8.** Em diversas aplicações matemáticas é necessário determinar o Máximo Divisor Comum (MDC) e o Mínimo Múltiplo Comum (MMC) entre dois números.

#### Faça um programa que leia dois números inteiros positivos e calcule inicialmente o MDC utilizando o seguinte processo:

- • enquanto os dois valores forem diferentes, subtraia o menor valor do maior;
- • quando os dois valores se tornarem iguais, esse valor será o MDC.

#### Depois, utilize o MDC encontrado para calcular o MMC por meio da relação:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_8.png" 
    alt="Gráfico da análise"
    width="600"
>
</div>

In [ ]:
def mdc_e_mmc(valor_a,valor_b):
    calculo_mdc= 0
    num1 = valor_a
    num2 = valor_b
    
    while num1 != num2:
        if num1 < num2:
            num2 = num2 - num1
        else:
            num1 = num1 - num2
    calculo_mdc = num2
    formula_mmc = valor_a * valor_b/ calculo_mdc
    print(f"Valor do Máximo Divisor Comum: {calculo_mdc}\nValor do MMC: {formula_mmc}")
    
        

In [5]:
valor_a = int(input("Digite o valor a: "))
valor_b = int(input("Digite o valor b: "))
mdc_e_mmc(valor_a, valor_b)


Valor do Máximo Divisor Comum:6
Valor do MMC:72.0


#### **9.** Um número inteiro positivo é chamado de primo quando possui exatamente dois divisores positivos: 1 e ele próprio. Faça um programa que leia um número inteiro e determine se ele é primo. O programa deve testar possíveis divisores utilizando uma estrutura de repetição.

#### Considere que números menores que 2 não são primos.

- Exemplos:
  - Entrada:
  - 17
  - Saida:
  - Numero primo

  - Entrada:
  - 21
  - Saida:
  - Numero nao primo

In [33]:
entrada = int(input("Digite o número: "))
num_primos = []
count_primos = 0
for i in range(1, entrada+1):
    if entrada%i ==0:
        count_primos +=1
        num_primos.append(i)
    else:
        continue
if count_primos ==2:
    print(f"Número é primo!\nOs divisores são 1 e ele mesmo\n{num_primos}")
else:
    print("Número não primo")

Número não primo


In [55]:



def eh_primo(num):
    count_primos = 0
    count = 0
    while count <=num:
        count +=1
        if num%count ==0:
            count_primos +=1
        else:
            continue
    if count_primos ==2:
        print(f'Número Primo\nNúmero: {num}\nQuantidade Divisores: {count_primos}')
    else:
        print(f"Número não Primo\nNúmero:{num}\nQuantidade Divisores: {count_primos}")

In [56]:
num = int(input("Digite um número maior que zero: "))
while num < 0:
    print("Maior que zero!")
    num = int(input("Digite um número maior que zero: "))
eh_primo(num)

Número não Primo
Número:6
Quantidade Divisores: 4


#### **10.** Alguns números possuem uma propriedade matemática particular: a soma de todos os seus divisores positivos, excluindo o próprio número, é igual ao número original.

#### Esses valores são chamados de números perfeitos.

#### Por exemplo, o número 6 possui como divisores próprios 1, 2 e 3:

- 1 + 2 + 3 = 6

#### Da mesma forma, os divisores próprios de 28 são 1, 2, 4, 7 e 14:

- 1 + 2 + 4 + 7 + 14 = 28
#### Faça um programa que leia um número inteiro positivo, calcule a soma de seus divisores próprios e informe se o número é perfeito.

#### Ao final, exiba também a soma encontrada.

#### Exemplo:

 -Entrada:
- 28
- Saida:
  - Soma dos divisores: 28
  - Numero perfeito

In [1]:
0%2

0

In [39]:
def divisores_perfeitos(num):
    num_original = num
    count = 1
    soma_dividor = 0
    while num != count:
        if num % count ==0:
            soma_dividor +=count
        print(count)
        count +=1
    if num_original == soma_dividor:
        print(f"É um número perfeito!\nNúmero Original {num_original} == Soma dividor:{soma_dividor}")
    else:
        print(f"Não é perfeito soma dos divisores:\n{soma_dividor} != {num_original}")
    

In [45]:
num = int(input("Digite um número para conferir a propriedade matemática particular:\na soma de todos os seus divisores positivos, excluindo o próprio número, é igual ao número original "))
while num <0:
    print("Apenas maiores que zero")
    num = int(input("Digite um número para conferir a propriedade matemática particular:\na soma de todos os seus divisores positivos, excluindo o próprio número, é igual ao número original "))
divisores_perfeitos(num)

Não é perfeito soma dos divisores:
0 != 1


#### **11.** Um número de três algarismos possui uma propriedade especial quando a soma do cubo de cada um de seus algarismos é igual ao próprio número.

#### Por exemplo:
- 153 = 13 + 53 + 33

#### Faça um programa que leia um número inteiro entre 100 e 999.

#### Sem converter o número para string, obtenha separadamente os algarismos das centenas, dezenas e unidades utilizando os operadores de divisão inteira (//) e resto da divisão (%).

#### Em seguida, calcule a soma dos cubos dos três algarismos e informe se o número possui a propriedade descrita.

#### Exemplos:

In [23]:
def possui_propriedade(num):
    num_entrada = num
    num_propriedade  = 0
    while num !=0:
        print(num)
        resto = num%10
        num_propriedade += resto **3
        num = num//10
        print(num)
    
    if num_propriedade == num_entrada:
        print(f"Soma: {num_propriedade}\nPossui Propriedade!")
    else:
        print(f"{num_propriedade}\nNão possui propriedade")

In [15]:
num = 153
resto = num%10
print(resto)
print(num//10)

3
15


In [24]:
num = int(input("Digite um número entre 100 e 999:"))
while num < 100 or num > 999:
    print("Apenas entre 0 e 999")
    num = int(input("Digite um número entre 100 e 999:"))
possui_propriedade(num)

153
15
15
1
1
0
Soma: 153
Possui Propriedade!


#### 12. Para determinados anos, a data do domingo de Páscoa pode ser calculada utilizando uma sequência de operações matemáticas.

#### Considere apenas anos entre 1900 e 2099. Leia o ano e realize os seguintes cálculos:

- a = ano mod 19
- b = ano mod 4
- c = ano mod 7
- d = (19a + 24) mod 30
- e = (2b + 4c + 6d + 5) mod 7

#### **Em seguida, calcule:**

#### dia = 22 + d + e

#### Para os anos 1954, 1981, 2049 e 2076, deve-se subtrair 7 do valor calculado para dia.

#### O valor obtido inicialmente representa uma data a partir de 22 de março. Portanto:

    - • se dia for menor ou igual a 31, a Páscoa ocorrerá em março;
    - • se dia for maior que 31, a Páscoa ocorrerá em abril, no dia dia - 31.

#### Faça um programa que leia o ano, execute os cálculos e exiba a data da Páscoa no formato dia/mes.

#### Caso o ano informado esteja fora do intervalo de 1900 a 2099, exiba uma mensagem informando que o ano é inválido.

In [64]:
def determinar_pascoa(ano):
    a = ano%19
    b = ano%4
    c = ano%7
    d = (19 * a + 24)%30
    e = (2*b + 4*c + 6*d +5)%7
    dia = 22 + d + e
    if ano in (1954,1981,2049,2076):
        dia-=7
        print(dia)
    if dia <=31:
        print(f"O dia {dia} de Páscoa ocorrerá em Março! Dado o ano {ano}")
    elif dia > 31:
        print(dia)
        dia-=31
        print(f"O dia {dia} de Páscoa ocorreá em Abril! Dado o ano {ano}")
    

In [71]:
x = -4
x = x + x + x
print(x)

-12


In [66]:
ano = int(input("Digite um ano no intervaldo de **1900-2099** para se calcular quando cairá o dia de páscoa!"))
while ano < 1900 or ano > 2099:
    print(f"O ano {ano} está fora do intervalo estabelecido para cálculos!")
    ano = int(input("Digite um ano no intervaldo de **1900-2099** para se calcular quando cairá o dia de páscoa!"))
determinar_pascoa(ano)


46
O dia 15 de Páscoa ocorreá em Abril! Dado o ano 1900


#### **13.** Um sistema recebe um número escrito em representação binária, composto apenas pelos algarismos 0 e 1, e precisa convertê-lo para sua representação decimal. Faça um programa que leia um número binário como um número inteiro e realize a conversão para decimal sem utilizar funções prontas de conversão de base.

#### O processamento deve considerar os algarismos da direita para a esquerda. Para cada algarismo, utilize a potência de 2 correspondente à sua posição.

#### Por exemplo:

- 11012 = 1 · 23 + 1 · 22 + 0 · 21 + 1 · 20 = 1310

#### O programa deve também verificar se o número informado é realmente binário. Caso algum algarismo diferente de 0 ou 1 seja encontrado, exiba:

#### Numero binario invalido

#### Caso contrário, exiba o valor decimal correspondente.
#### Não utilize conversão do número para string.

In [24]:
numero = 100001
count = 0
while numero !=0:
    resto = numero%10
    print(resto)
    numero = numero//10
    print(numero)
    count+=1
print(count)

1
10000
0
1000
0
100
0
10
0
1
1
0
6


In [46]:
def converte_binario(entrada):
    soma_numeros = 0
    convertido = 1
    count = 0
    while entrada!=0:
        resto = entrada%10
        print(resto)
        convertido = resto * (2 ** count)
        soma_numeros+= convertido
        count +=1
        entrada = entrada//10
        print(entrada)
    
    return soma_numeros

In [20]:
entrada = 1000000000001
print(converte_binario(entrada))

1
100000000000
0
10000000000
0
1000000000
0
100000000
0
10000000
0
1000000
0
100000
0
10000
0
1000
0
100
0
10
0
1
1
0
(4097, 13, 4096)


In [23]:

converter_numero_binario = input("Deseja converter um número para Binário?").lower()
while converter_numero_binario not in("sim", "não", "nao", "s", "n"):
    print("Digite apenas as opções:\n sim, não, nao, s, n")
    converter_numero_binario = input("Deseja converter um número para Binário?").lower()
while converter_numero_binario not in ("nao", "não", "n"):
    numero_bin = input("Digite um número: ")
    caractere_numero = str(numero_bin)
    binario = True
    while True:
        for i in caractere_numero:
            if i in('23456789'):
                print(f"O número: {numero_bin} não é binário. pois {i} não é binário ")
                binario = False
            else:
                binario = True
        if binario == False:
            numero_bin = input("Digite um número: ")
            caractere_numero = str(numero_bin)
        elif binario:
            print("Número Binário!")
            numero_bin = int(caractere_numero)
            print(numero_bin)
            break
    print(f"O número binário convertido é: {converte_binario(numero_bin)}")
    converter_numero_binario = input("Deseja converter um número para Binário?").lower()
    while converter_numero_binario not in("sim", "não", "nao", "s", "n"):
        print("Digite apenas as opções:\n sim, não, nao, s, n")
        converter_numero_binario = input("Deseja converter um número para Binário?").lower()





Digite apenas as opções:
 sim, não, nao, s, n
O número: 282282 não é binário. pois 2 não é binário 
O número: 282282 não é binário. pois 8 não é binário 
O número: 282282 não é binário. pois 2 não é binário 
O número: 282282 não é binário. pois 2 não é binário 
O número: 282282 não é binário. pois 8 não é binário 
O número: 282282 não é binário. pois 2 não é binário 
Número Binário!
101010101010101010101
1
10101010101010101010
0
1010101010101010101
1
101010101010101010
0
10101010101010101
1
1010101010101010
0
101010101010101
1
10101010101010
0
1010101010101
1
101010101010
0
10101010101
1
1010101010
0
101010101
1
10101010
0
1010101
1
101010
0
10101
1
1010
0
101
1
10
0
1
1
0
O número binário convertido é: 1398101


#### 19

In [ ]:
def suporte_chamados(qtd_dias, qtd_chamados):
    maior_trecho = 0
    trecho_atual_maior = 0
    primeiro_dia = 0
    ultimo_dia = 0
    for i in range(1, qtd_dias+1):
        chamados = int(input("Digite o primeiro chamado: "))
        primeiro_dia = cha

#### Determine a saída produzida para cada conjunto de entrada:
a)-4 - 8
b)- 5 - 2

#### 14. Peça ao usuário que informe uma sequência de números inteiros. A leitura deve ser encerrada quando o valor 0 for digitado. O valor 0 serve apenas como sentinela e não deve fazer parte da sequência. Um trecho crescente consecutivo é formado por números em que cada elemento é estritamente maior que o elemento imediatamente anterior.

#### Desenvolva um programa que determine o tamanho do maior trecho crescente consecutivo existente na sequência.

#### Por exemplo:
#### Entrada:
- 3
- 5
- 7
- 2
- 4
- 6
- 8
- 1
- 0

#### Saida:
#### Maior trecho crescente: 4
#### Nesse exemplo, o maior trecho é formado por:

- 2, 4, 6, 8

#### Caso apenas um número seja informado antes da sentinela, o maior trecho terá tamanho 1.

In [29]:
def leitura_ciclo(entrada):
    atual_sequencia = 1
    maior_sequencia = 1
    anterior = entrada
    while entrada !=0:
        print("Digite o próximo número da sequência: ")
        entrada = int(input("Digite o próximo número da sequência!"))
        print(f"Número Digitado: {entrada} anterior: {anterior}")
        if entrada > anterior:
            atual_sequencia +=1
        else:
            atual_sequencia =1
        if atual_sequencia > maior_sequencia:
            maior_sequencia = atual_sequencia
        anterior = entrada
    print(f"Maior trecho da sequência: {maior_sequencia}")

In [30]:
entrada = int(input("Digite o primeiro número da sequência!"))
print(leitura_ciclo(entrada))

Digite o próximo número da sequência: 
Número Digitado: 5 anterior: 3
Digite o próximo número da sequência: 
Número Digitado: 7 anterior: 5
Digite o próximo número da sequência: 
Número Digitado: 2 anterior: 7
Digite o próximo número da sequência: 
Número Digitado: 4 anterior: 2
Digite o próximo número da sequência: 
Número Digitado: 6 anterior: 4
Digite o próximo número da sequência: 
Número Digitado: 8 anterior: 6
Digite o próximo número da sequência: 
Número Digitado: 1 anterior: 8
Digite o próximo número da sequência: 
Número Digitado: 0 anterior: 1
Maior trecho da sequência: 4
None


#### **15.** Um equipamento registra continuamente códigos numéricos inteiros. A leitura termina quando o valor 0 é informado.

#### Deseja-se identificar repetições consecutivas de um mesmo código. Desenvolva um programa que determine:
- • o maior número de repetições consecutivas encontrado;
- • qual foi o número que formou essa maior sequência.

#### Por exemplo, para a sequência:

- 4, 4, 7, 7, 7, 2, 2, 5, 5, 5, 5, 3, 0 o resultado deve ser:

- Numero: 5
- Maior sequencia: 4

#### Se houver empate entre duas sequências de mesmo tamanho, considere aquela que apareceu primeiro.

In [32]:
def sequencia_iguais(entrada):
    atual_sequencia = 1
    maior_sequencia = 1
    numero_maior_sequencia = entrada
    anterior = entrada
    while entrada !=0:
        print("Digite o próximo número da sequência: ")
        entrada = int(input("Digite o próximo número da sequência!"))
        print(f"Número Digitado: {entrada} anterior: {anterior}")
        if entrada == anterior:
            atual_sequencia +=1
        else:
            atual_sequencia =1
        if atual_sequencia > maior_sequencia:
            maior_sequencia = atual_sequencia
            numero_maior_sequencia = entrada
        anterior = entrada
    print(f"Maior trecho da sequência de números repetidos:{maior_sequencia}\nNúmero que se repete: {numero_maior_sequencia}")


In [33]:
entrada = int(input("Digite o primeiro número da sequência!"))
print(sequencia_iguais(entrada))

Digite o próximo número da sequência: 
Número Digitado: 4 anterior: 4
Digite o próximo número da sequência: 
Número Digitado: 7 anterior: 4
Digite o próximo número da sequência: 
Número Digitado: 7 anterior: 7
Digite o próximo número da sequência: 
Número Digitado: 7 anterior: 7
Digite o próximo número da sequência: 
Número Digitado: 2 anterior: 7
Digite o próximo número da sequência: 
Número Digitado: 2 anterior: 2
Digite o próximo número da sequência: 
Número Digitado: 5 anterior: 2
Digite o próximo número da sequência: 
Número Digitado: 5 anterior: 5
Digite o próximo número da sequência: 
Número Digitado: 5 anterior: 5
Digite o próximo número da sequência: 
Número Digitado: 5 anterior: 5
Digite o próximo número da sequência: 
Número Digitado: 3 anterior: 5
Digite o próximo número da sequência: 
Número Digitado: 0 anterior: 3
Maior trecho da sequência de números repetidos:4
Número que se repete: 5
None


#### **16** Considere a seguinte série:

<div align="center">
<img 
    src="print_abertas_lista_1_02/quetao_16.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>


#### e assim sucessivamente.

#### Faça um programa que leia um número inteiro positivo n e calcule:
- • cada um dos n primeiros termos da série;
- • a soma final desses termos.

#### O programa deve aproveitar o produto calculado no termo anterior, evitando recalcular todo o produtório desde o início a cada repetição.

In [90]:
def soma_series(num):
    soma_serie = 0
    numerador = 1
    denominador = 0
    anterior = 0
    for i in range(1, num+1):
        anterior   = i+1
        numerador *=anterior
        print(f'Rodada: {i}\nNumerador: {numerador}')
        denominador = i **2
        print(f'Rodada: {i}\nDenominador: {denominador}')
        n_termo = numerador/denominador
        print(f"O {i} termo é: {n_termo}")
        soma_serie += n_termo
    print(f"Soma série: {soma_serie}")


In [91]:
num = int(input("Digite um número: "))
while num <0: 
    print("Maior que zero!")
    num = int(input("Digite um número: "))
soma_series(num)

Rodada: 1
Numerador: 2
Rodada: 1
Denominador: 1
O 1 termo é: 2.0
Rodada: 2
Numerador: 6
Rodada: 2
Denominador: 4
O 2 termo é: 1.5
Rodada: 3
Numerador: 24
Rodada: 3
Denominador: 9
O 3 termo é: 2.6666666666666665
Rodada: 4
Numerador: 120
Rodada: 4
Denominador: 16
O 4 termo é: 7.5
Rodada: 5
Numerador: 720
Rodada: 5
Denominador: 25
O 5 termo é: 28.8
Rodada: 6
Numerador: 5040
Rodada: 6
Denominador: 36
O 6 termo é: 140.0
Soma série: 182.46666666666667


#### **17**:

<div align="center">
<img 
    src="print_abertas_lista_1_02/quetao_17.png" 
    alt="Gráfico da análise"
    width="900"
>
</div>




In [11]:

for i in range(1,6):
    if i%2 == 0:
        print(i, "é par")
    elif i%2 !=0:
        print(i, "é ímpar")

1 é ímpar
2 é par
3 é ímpar
4 é par
5 é ímpar


In [38]:
def series_complexas(num):
    numerador = 1
    denominador = 1
    soma_serie = 0
    anterior = 1
    for i in range(1, num+1):
        if i%2 !=0:
            numerador  *=i
        print(f"Rodada:{i}")
        print(f'numerador:{numerador}')
        denominador = 2**i * i
        print(f"denominador: {denominador}")
        n_termo = numerador/denominador
        soma_serie += n_termo
        print(f"{i}º Termo: {n_termo}")
    
       
    return soma_serie

In [37]:
num = int(input("Digite um número: "))
while num <0: 
    print("Maior que zero!")
    num = int(input("Digite um número: "))
series_complexas(num)

1
Rodada:1
numerador:1
denominador: 2
1º Termo: 0.5
1
2
Rodada:2
numerador:1
denominador: 8
2º Termo: 0.125
1
2
3
Rodada:3
numerador:3
denominador: 24
3º Termo: 0.125
1
2
3
4
Rodada:4
numerador:9
denominador: 64
4º Termo: 0.140625
1
2
3
4
5
Rodada:5
numerador:135
denominador: 160
5º Termo: 0.84375


1.734375

#### **18**



In [2]:
def silo_estoque(qtd_inicial):
    qtd_agricultores  =0
    qtd_sementes  =0
    qtd_restante = qtd_inicial
    while True:
        retirada = float(input(f"Digite a quantidade que se deseja retirar do silo: {qtd_inicial}"))
        while retirada < 0:
            print("Não existe retirada negativa!")
            retirada = float(input(f"Digite a quantidade que se deseja retirar do silo: {qtd_inicial}"))
        if retirada <= qtd_inicial:
            qtd_inicial -= retirada
            qtd_agricultores +=1
            qtd_sementes += retirada
            qtd_restante -= retirada
            continue
        elif retirada > qtd_inicial:
            print(f"Pedido {retirada} maior que a capacidade do silo inicial: {qtd_inicial}")
            break
    if qtd_agricultores ==0:
        print(f"Nenhum agricultor foi atendido!Já que o silo está zerado")
    else:
        print(f"Total de Agricultores Atendidos: {qtd_agricultores}\nTotal de Sementes retiradas: {qtd_sementes}\nQuantidade restante do silo:{qtd_restante}")


In [1]:
quantidade_silo = float(input("Digite a quantidade inicial do silo!"))
while quantidade_silo < 0:
    print("Não existe estoque negativo!")
    quantidade_silo = float(input("Digite a quantidade inicial do silo!"))
silo_estoque(quantidade_silo)


NameError: name 'silo_estoque' is not defined

#### 19. Uma empresa acompanha diariamente a quantidade de chamados de suporte recebidos.

#### Para analisar períodos de maior demanda, deseja-se identificar o maior trecho de dias consecutivos em que a quantidade de chamados aumentou em relação ao dia anterior.

#### Faça um programa que leia inicialmente a quantidade N de dias analisados e, em seguida, leia a quantidade de chamados de cada um desses dias.
#### O programa deve determinar:

- • o tamanho do maior período consecutivo de crescimento;
- • a quantidade de chamados registrada no primeiro dia desse período;
- • a quantidade de chamados registrada no último dia desse período.

#### Por exemplo, considerando:

- 10, 14, 18, 8, 9, 15, 21, 25, 7
  - o maior período crescente é:
  - 8, 9, 15, 21, 25 e possui tamanho 5.

#### A saída correspondente seria:

#### Maior periodo: 5
#### Inicio: 8
#### Fim: 25

#### Considere que N ≥ 1.

In [31]:
def suporte_chamados(num_chamados):
    atual_maior = 0
    maior_tamanho = 1
    anterior = float("-inf")
    primeiro_dia = 0
    ultimo_dia = 0
    inicio_maior = 0
    fim_maior = 0
    for i in range(1,num_chamados+1):
        chamados = int(input("Digite o valor do chamado: "))
        print(f"Valor i > {i}/Chamado: {chamados}")
        if anterior < chamados:
            primeiro_dia = anterior
            maior_tamanho +=1
            if maior_tamanho > atual_maior:
                atual_maior = maior_tamanho
            if ultimo_dia < chamados:
                ultimo_dia = chamados
            if primeiro_dia  > chamados:
                primeiro_dia = chamados
        else:
            inicio_maior = primeiro_dia
            fim_maior = ultimo_dia
            maior_tamanho =1
        anterior = chamados
    print(f"Maior Período: {atual_maior}\nInício: {inicio_maior}\nÚltimo Dia: {fim_maior}")
        

In [30]:
num_chamados = int(input("Digite o número de chamados, tal que N>=1"))
while num_chamados < 1:
    print("Não pode ser menor que 1")
    num_chamados = int(input("Digite o número de chamados, tal que N>=1"))
suporte_chamados(num_chamados)



Valor i > 1/Chamado: 10
Valor i > 2/Chamado: 14
Valor i > 3/Chamado: 18
Valor i > 4/Chamado: 8
Valor i > 5/Chamado: 9
Valor i > 6/Chamado: 15
Valor i > 7/Chamado: 21
Valor i > 8/Chamado: 25
Valor i > 9/Chamado: 7
Maior Período: 5
Início: 21
Último Dia: 25


#### **20.** Um sistema precisa gerar um padrão numérico a partir de um valor inteiro positivo N.

#### Primeiramente, devem ser impressas linhas crescentes. A linha i deve apresentar os números de 1 até i.
#### Depois de atingir N, o padrão deve diminuir novamente até 1.
#### Por exemplo, para N = 5:
- 1
- 12
- 123
- 1234
- 12345
- 1234
- 123
- 12
- 1

#### Entretanto, o programa deve seguir uma restrição adicional: cada linha deve serconstruída por meio de uma repetição, imprimindo ou concatenando individualmente os valores que fazem parte dela.
#### Faça o programa utilizando estruturas de repetição aninhadas.

In [13]:
def exibe_valores(num):
    for i in range(1,5+1):
        print(i)
        for j in range(num, i+1,1):
            print(j)

In [14]:
num = int(input("Digite um número maior que 0: "))
while num < 0:
    print("O número deve ser maior que 0!")
    num = int(input("Digite um número maior que 0: "))
exibe_valores(num)

1
2
3
4
5
5


#### **21** Um laboratório de controle ambiental acompanha diariamente a quantidade de microalgas presentes em um reservatório experimental. No início do experimento, o reservatório possui uma população conhecida. A cada dia, a população sofre algumas alterações sempre na mesma ordem. Primeiro, a quantidade de microalgas dobra e, em seguida, três unidades são perdidas naturalmente. A cada quarto dia do experimento, são adicionadas dez unidades de nutrientes, aumentando imediatamente a população em dez. Além disso, a cada sexto dia é realizada uma filtragem que remove metade da população existente naquele momento; caso a quantidade seja ímpar, deve ser considerada apenas a parte inteira da metade. Em alguns dias, mais de uma intervenção pode ocorrer. No décimo segundo dia, por exemplo, a adição de nutrientes ocorre antes da filtragem, pois as regras devem ser aplicadas na ordem em que foram descritas.

#### Escreva um programa que leia a população inicial de microalgas, a quantidade de dias do experimento e um limite de segurança. O programa deve simular todos os dias e, ao final, informar a população existente, a maior população registrada ao final de um dia e a quantidade de dias em que a população terminou estritamente acima do limite de segurança.

In [2]:
def controle_ambiental(pop_inicial, qtd_dias, limite_seguranca):
    maior_populacao = pop_inicial
    qtd_dias_acima = 0
    pop_existente = pop_inicial
    for i in range(1, qtd_dias+1):
        pop_inicial = pop_inicial * 2 -3
        if pop_inicial >limite_seguranca:
            qtd_dias_acima +=1
        if pop_inicial > maior_populacao:
            maior_populacao = pop_inicial
        if i %4 ==0:
            pop_inicial = pop_inicial +10
            if pop_inicial >limite_seguranca:
                qtd_dias_acima +=1
            if pop_inicial > maior_populacao:
                maior_populacao = pop_inicial
        elif i%6 ==0:
            if pop_inicial %2 !=0 or pop_inicial %2 ==0:
                pop_inicial = pop_inicial//2
                if pop_inicial > limite_seguranca:
                    qtd_dias_acima +=1
                if pop_inicial > maior_populacao:
                    maior_populacao = pop_inicial
    print(f"População Inicial: {pop_existente}\nPopulação restante: {pop_inicial}\nQuantidade de Microalgas acima do limite de segurança:{qtd_dias_acima} ")
           

In [ ]:
pop_inicial = int(input("Informe a população Inicial: "))
while pop_inicial < 0:
    print("Não existe população negativa!")
    pop_inicial = int(input("Informe a população Inicial: "))
qtd_dias  = int(input("Informe a quantidade de dias do experimento: "))
while qtd_dias < 0:
    print("Não existe dias negativos!")
    qtd_dias  = int(input("Informe a quantidade de dias do experimento: "))
limite_seguranca = int(input("Informe a quantidade de microalgas para o limite de segurança: "))
while limite_seguranca < 0:
    print("Não existe limites negativos!")
    limite_seguranca = int(input("Informe a quantidade de microalgas para o limite de segurança: "))
controle_ambiental(pop_inicial, qtd_dias, limite_seguranca)

ValueError: invalid literal for int() with base 10: 'e'

#### **22.** Um sistema de autenticação utiliza frases secretas cuja sequência de letras deve apresentar uma propriedade de simetria. Para verificar uma frase, os espaços existentes nela são ignorados. Assim, a frase apos a sopa é considerada válida porque, depois da remoção dos espaços, obtém-se aposasopa, que pode ser lida da mesma maneira da esquerda para a direita e da direita para a esquerda.

#### Escreva um programa que leia uma frase formada apenas por letras minúsculas e espaços e determine se ela atende à regra de autenticação. Durante o processamento, o programa deve construir uma nova string sem os espaços e outra string com os mesmos caracteres na ordem inversa. A inversão deve ser construída por meio de uma estrutura de repetição, sem utilizar fatiamento como texto[::-1]. Ao final, o programa deve informar se a frase é válida e quantas letras foram consideradas na verificação.

In [ ]:
entrada = "asjehrf"
type(entrada)


str

In [23]:
len(entrada)

7

In [34]:
def minha_len(entrada):
    count_palavras = 0
    count_numeros = 0
    if type(entrada) == str:
        for i in entrada:
            count_palavras +=1
        return count_palavras
    else:
        for i in range(entrada):
            count_numeros+=1
        return count_numeros


In [35]:
minha_len(entrada)

7

In [43]:
letra  ="sopa"
nova_letra= ""
for i in range(len(letra)-1,-1,-1):
    print(letra[i])
    nova_letra += letra[i]
print(nova_letra)

a
p
o
s
apos


In [50]:
def eh_palindromo_ao_contrario(entrada):
    count_letras_validas = 0
    ao_contrario = ""
    nova_palavra = ""
    palindromo = True
    for i in range(minha_len(entrada)):
        if entrada[i] ==" ":
            continue
        else:
            nova_palavra += entrada[i]
            count_letras_validas +=1
    for i in range(len(nova_palavra)-1,-1,-1):
        ao_contrario += nova_palavra[i]
    if ao_contrario == nova_palavra:
        palindromo = True
        if palindromo:
            print(f"É palíndromo!\nPalavra sem espaços: {nova_palavra} ao contrário: {ao_contrario}\nQuantidade de letras: {count_letras_validas}")
    elif ao_contrario != nova_palavra:
        palindromo = False
        if palindromo == False:
            print(f"NÃO é palíndromo!\nPalavra sem espaços:{nova_palavra} ao contrário: {ao_contrario}")

In [52]:
entrada = input("Digite uma palavra para saber se é palíndromo ou não!").lower()
eh_palindromo_ao_contrario(entrada)

É palíndromo!
Palavra sem espaços: asacadadacasa ao contrário: asacadadacasa
Quantidade de letras: 13


#### **23.** Um robô de inspeção precisa deslocar-se em um galpão organizado como uma grade retangular. Para sair de sua posição inicial e alcançar o destino pelo menor caminho possível, o robô deverá realizar exatamente D movimentos para a direita e B movimentos para baixo, mas esses movimentos podem ocorrer em diferentes ordens. Se forem necessários dois movimentos para a direita e um movimento para baixo, por exemplo, existem três caminhos mínimos possíveis: DDB, DBD e BDD. A quantidade de caminhos mínimos distintos pode ser calculada por

#### **C = (D + B)!/ D! · B!** .

#### Escreva um programa que leia os valores inteiros não negativos D e B e determine a quantidade de caminhos mínimos diferentes que o robô pode realizar. Os três fatoriais necessários devem ser calculados manualmente por meio de estruturas de repetição, sem utilizar math.factorial()

In [55]:
def fatorial(num):
    if num ==0:
        return 1
    elif num!=0:
        fatorial = 1
        for i in range(1, num+1):
            fatorial *=i
            print(i)
        return fatorial

In [57]:
num = 5
fatorial(num)

1
2
3
4
5


120

In [64]:
fatorial(12)

1
2
3
4
5
6
7
8
9
10
11
12


479001600

In [74]:
def caminhos_robo(valor_d, valor_b):
    soma = valor_b + valor_d
    caminhos =  fatorial(soma)/ (fatorial(valor_d) * fatorial(valor_b))
    return caminhos

In [77]:
valor_d = int(input("Digite o valor d: "))
while valor_d < 0:
    print("Deve ser maior que 0!")
    valor_d = int(input("Digite o valor d: "))
valor_b = int(input("Digite o valor b: "))
while valor_b < 0:
    print("Deve ser maior que 0!")
    valor_b = int(input("Digite o valor b: "))
print(caminhos_robo(valor_d, valor_b))




1
2
3
4
5
6
7
8
9
10
11
12
1
2
3
4
5
6
1
2
3
4
5
6
924.0


#### **24.** Uma equipe responsável pela infraestrutura de um serviço observou um padrão no crescimento da quantidade de requisições processadas por determinado módulo.

#### No primeiro intervalo foi registrada uma requisição e, no segundo, também uma requisição. A partir do terceiro intervalo, a quantidade registrada é sempre igual à soma das quantidades observadas nos dois intervalos imediatamente anteriores. Dessa forma, os primeiros valores são

- 1, 1, 2, 3, 5, 8, 13, 21, . . .

#### Escreva um programa que leia um número inteiro positivo N, representando a quantidade de intervalos que devem ser analisados, e a capacidade máxima de requisições que o servidor consegue processar em um único intervalo. O programa deve gerar os N primeiros valores dessa sequência e informar o valor correspondente ao N-ésimo intervalo, a soma das requisições registradas em todos os intervalos e o primeiro intervalo em que a quantidade de requisições ultrapassou a capacidade do servidor.

#### Caso isso não aconteça nos N intervalos analisados, o programa deve informar que a capacidade não foi ultrapassada. Os casos N = 1 e N = 2 também devem ser tratados corretamente.

In [19]:
def infraestrutura_servidor(qtd_intervalos, num_requisicoes):
    soma_intervalo= 0
    intervalos_limites = ""
    count = 0
    count_valores =0
    for i in range(1, qtd_intervalos+1):
        count +=1
        count_valores += i
        print(f"Intervalo: {i}: count valores: {count_valores}")
        soma_intervalo +=count_valores
        if soma_intervalo > num_requisicoes:
            intervalos_limites = i
    if intervalos_limites == "" or count <=2:
        print("Não ultrapassou a capacidade de requisições do servidores")
    else:
        print(f"Soma dos intervalos: {soma_intervalo}\nIntervalo Limite: {i}")

In [20]:
quantidade_intervalos = int(input("Digite a quantidade de Intervalos: "))
while quantidade_intervalos < 0:
    print("Deve ser maior que zero")
    quantidade_intervalos = int(input("Digite a quantidade de Intervalos: "))
capacidade_requisicoes = int(input("Digite a capacidade máxima de requisições: "))
while capacidade_requisicoes < 0:
    print("Deve ser maior que zero!")
    capacidade_requisicoes = int(input("Digite a capacidade máxima de requisições: "))
infraestrutura_servidor(quantidade_intervalos, capacidade_requisicoes)

    


Intervalo: 1: count valores: 1
Intervalo: 2: count valores: 3
Intervalo: 3: count valores: 6
Intervalo: 4: count valores: 10
Intervalo: 5: count valores: 15
Soma dos intervalos: 35
Intervalo Limite: 5


#### **25.** Uma instituição utiliza códigos numéricos de nove algarismos para identificar documentos internos. Os oito primeiros algarismos representam o número do documento e o último é um dígito verificador utilizado para detectar erros de digitação. Para calcular esse dígito, os oito primeiros algarismos são multiplicados, da esquerda para a direita, pelos pesos 2, 3, 4, 5, 6, 7, 8 e 9. Os produtos obtidos são somados, produzindo um valor S, e então calcula-se

#### R = S mod 11.

#### Quando R for igual a 0 ou 1, o dígito verificador esperado será 0. Nos demais casos, ele será calculado por 11 − R. Caso esse cálculo resulte em 10, o dígito também deverá ser considerado 0.

#### Escreva um programa que leia o código completo como uma string contendo exatamente nove algarismos, calcule o dígito verificador esperado e compare-o com o último algarismo informado. Ao final, exiba o dígito calculado e informe se o código é válido ou inválido. O código completo não deve ser convertido para um número inteiro.

In [ ]:
def codigos_numericos_verificador(entrada):
    ultimo_algarismo = int(entrada[-1])
    entrada_sem_o_ultimo = entrada[:len(entrada)-1]
    valor_1= 1
    codigo_s = 0
    for i in range(len(entrada_sem_o_ultimo)):
        soma_s = 0
        valor_1 +=1
        if entrada_sem_o_ultimo[i] == "0" or entrada_sem_o_ultimo[i] =="1" or entrada_sem_o_ultimo[i] =="2" or entrada_sem_o_ultimo[i] =="3" or entrada_sem_o_ultimo[i] =="4" or entrada_sem_o_ultimo[i]=="5" or entrada_sem_o_ultimo[i] == "6" or entrada_sem_o_ultimo[i] =="7" or entrada_sem_o_ultimo[i] =="8" or entrada_sem_o_ultimo[i] =="9":
            valor = int(entrada_sem_o_ultimo[i])
            soma_s = valor * valor_1
            codigo_s += soma_s
    calculo_r = codigo_s%11
    if calculo_r ==0 or calculo_r ==1:
        esperado = 0
        print("O ÚLTIMO digito esperado é 0!")
        if int(ultimo_algarismo) == esperado:
            print(f'Código digitado: {entrada} possui como último algarismo: {ultimo_algarismo} que é igual ao esperado: {esperado}')
    else:
        esperado = 11 - calculo_r
        if esperado == int(ultimo_algarismo):
                print(f"É igual ao último algarismo!\nCódigo:{entrada} Esperado: {esperado}último algarismo: {ultimo_algarismo}")
        elif calculo_r ==10:
            esperado  = 0
            print(f"Código:{entrada} possui como algarismo {ultimo_algarismo}que é igual ao  esperado que é 0: {esperado}")
        else:
            print("Não é igual ao esperado!")
        
            

27. Uma empresa precisa transmitir pequenos códigos textuais por uma conexão de
baixa velocidade e deseja utilizar uma forma simples de compactação. Cada sequência
de caracteres iguais e consecutivos deve ser substituída pelo próprio caractere
seguido da quantidade de vezes que ele aparece naquele trecho. Assim, a mensagem
AAABBCCCCDAA deve ser transformada em A3B2C4D1A2. Observe que os dois
grupos da letra A devem ser tratados separadamente, pois não são consecutivos.
Escreva um programa que leia uma string não vazia e produza sua representação
compactada. Durante o processamento, determine também quantos grupos consecutivos
foram encontrados, qual foi o maior número de repetições consecutivas de um
mesmo caractere e qual caractere formou esse grupo. Em caso de empate, considere
o primeiro grupo encontrado. Ao final, compare os tamanhos do texto original e
do texto compactado e informe se a compactação realmente reduziu o tamanho da
mensagem.

In [51]:
def compactacao_mensagem(texto):
    novo_texto = ""
    count_grupos_consecutivos = 0
    maior_repeticao = float("-inf")
    caractere_maior = ""
    count_iguais =1
    for i in range(len(texto)-1):
        caractere_atual = texto[i]
        if caractere_atual == texto[i+1]:
            count_grupos_consecutivos+=1
            count_iguais +=1
            contagem_iguais = str(count_iguais)
        elif caractere_atual != texto[i+1]:
            novo_texto = texto[i]+contagem_iguais
            if count_iguais > maior_repeticao:
                maior_repeticao = count_iguais
                caractere_maior = texto[i]
            count_iguais =1
    
    tamanho_novo_texto =len(novo_texto)
    if count_grupos_consecutivos ==0:
        print("Nenhum caracterese consecutivo foi encontradao!")
    else:
        print(f"Número de caracteres consecutivos: {count_grupos_consecutivos}\nMaior número de repetições consecutivas:{maior_repeticao}\n Caractere que teve o maioir número de repetições consecutivas: {caractere_maior}")
        if tamanho_novo_texto < len(texto):
            print("Texto formado é menor!")
            print(f'\n Texto: {novo_texto}\n Tamanho: {tamanho_novo_texto}')
        else:
            print("A codificação não reduziu o tamanho do texto!")




In [52]:
mensagem = input("Informe a mensagem: ").lower()
while mensagem  in (" ", ""):
    print("Mensagem não pode ser vazia!")
    mensagem = input("Informe a mensagem: ")
compactacao_mensagem(mensagem)


Número de caracteres consecutivos: 7
Maior número de repetições consecutivas:4
 Caractere que teve o maioir número de repetições consecutivas: c
Texto formado é menor!

 Texto: d4
 Tamanho: 2


In [ ]:
texto = 'AAABBCCCCDAA'
count_vezes = 1
for i in range(len(texto)-1):
    count_vezes +=1
    if texto[i] == texto[i+1]:
        print(texto[i])
        print("é IGUAL!")
    else:
        print('Diferente')

A
é IGUAL!
A
é IGUAL!
Diferente
B
é IGUAL!
Diferente
C
é IGUAL!
C
é IGUAL!
C
é IGUAL!
Diferente
Diferente
A
é IGUAL!


In [46]:
len(texto)
print(count_vezes)


11


In [14]:
novo_texto = 'A3B2C4D1A2'
len(novo_texto)

10

#### **28.** Uma estação meteorológica registra uma sequência de temperaturas em intervalos regulares. Para analisar mudanças bruscas no comportamento observado, os pesquisadores desejam identificar os picos locais da série. Uma medição é consideradaum pico local quando sua temperatura é estritamente maior que a medição imediatamente anterior e também estritamente maior que a medição imediatamente posterior. Na sequência:

#### > 18, 21, 19, 22, 27, 25, 20,

#### por exemplo, as temperaturas 21 e 27 são picos locais. A primeira e a última medição nunca podem ser classificadas dessa forma, pois não possuem dois vizinhos. Escreva um programa que leia inicialmente um número inteiro N, com N ≥ 3, e depois as N temperaturas registradas. As medições devem ser processadas à medida que forem lidas, sem utilizar listas. Ao final, informe quantos picos locais foram encontrados e, caso exista pelo menos um, apresente também o valor do maior pico e sua posição, considerando que a primeira medição ocupa a posição 1. Se nenhum pico for encontrado, o programa deverá informar essa situação.

In [8]:
def pico_estacao_metereologica(n_temperaturas):
    maior_valor = float("-inf")
    picos_encontrados = 0
    for i in range(1, n_temperaturas+1):
        temperaturas = int(input(f"Digite a {i} temperatura: "))
        anterior = temperaturas
        i-=1
        atual = int(input(f"Digite a próxima {i}temperatura"))
        i-=1
        proxima = int(input(f"Digite a próxima {i} temperatura: "))
        if anterior < atual and atual > proxima:
            picos_encontrados +=1
            if atual > maior_valor or i  ==1:
                maior_valor = atual
    if picos_encontrados ==0:
        print("Nenhum pico foi encontrado!")
    else:
        print(f"Número de picos encontrados: {picos_encontrados}\nMaior valor no pico: {maior_valor}")

In [9]:
n_temperaturas = int(input("Digite a quantidadede temperaturas que será registrada: "))
while n_temperaturas < 3:
    print("A quantidade deverá ser maior que 3: ")
    n_temperaturas = int(input("Digite a quantidadede temperaturas que será registrada: "))
pico_estacao_metereologica(n_temperaturas)


Número de picos encontrados: 3
Maior valor no pico: 2992


In [3]:
for i in range(n_temperaturas):
    print(i)

0
1
2


#### **29.** Uma sonda espacial utiliza um procedimento numérico para reduzir códigos inteiros antes de transmiti-los para a estação de controle. O processo começa com um número inteiro positivo N. Sempre que o valor atual for par, ele é substituído por sua metade. Quando for ímpar, ele é substituído por três vezes seu valor acrescido de uma unidade. O procedimento é repetido até que o valor 1 seja alcançado. Por exemplo, começando em 6, são produzidos os valores

- 6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1.

#### Escreva um programa que leia o valor inicial e execute esse procedimento. Ao final, informe quantas transformações foram necessárias até chegar a 1, qual foi o maior valor alcançado durante todo o processo, quantas transformações foram realizadas quando o valor corrente era par e quantas foram realizadas quando ele era ímpar. O valor inicial também deve ser considerado na determinação do maior valor alcançado.

In [32]:
def sonda(entrada):
    count_transformacoes = 0
    valor = entrada
    maior_valor = entrada
    count_par = 0
    count_impar = 0
    while entrada !=1:
        if maior_valor < entrada:
            maior_valor = entrada
        if entrada %2 ==0:
            #count_transformacoes +=1
            count_par +=1
            entrada  = entrada //2
        elif entrada %2 !=0:
            count_impar +=1
            #count_transformacoes +=1
            entrada = (entrada *3) +1
            if maior_valor < entrada:
                maior_valor = entrada
        count_transformacoes +=1
    return (f'Transformações totais: {count_transformacoes}\nMaior Valor: {maior_valor}\nQuantidade de Pares: {count_par}\nQuantidade de ímpar: {count_impar}')

        

In [36]:
entrada = int(input("Digite o número: "))
sonda(entrada)

'Transformações totais: 7\nMaior Valor: 16\nQuantidade de Pares: 5\nQuantidade de ímpar: 2'

In [7]:
entrada = int(input("Digite o número: "))
sonda(entrada)

'Transformações totais: 8\nMaior Valor: 16\nQuantidade de Pares: 6\nQuantidade de ímpar: 2'

#### **30.** Um pequeno processador embarcado precisa realizar multiplicações, mas seu hardware não possui uma instrução específica para essa operação. Para contornar essa limitação, pode-se utilizar um algoritmo baseado apenas em somas, duplicações e divisões inteiras por 2. O processamento começa com dois valores positivos A e B.

#### Enquanto B for maior que zero, se B for ímpar, o valor atual de A é acrescentado ao resultado. Em seguida, A é duplicado e B é substituído por sua divisão inteira por 2.

#### Por exemplo, para calcular 13 × 11, os pares de valores processados são (13, 11), (26, 5), (52, 2) e (104, 1). Como os valores 11, 5 e 1 são ímpares, o resultado é obtido pela soma 13 + 26 + 104 = 143.

#### Escreva um programa que leia dois números inteiros, que podem ser positivos, negativos ou zero, e calcule seu produto utilizando esse procedimento. O programa deve tratar adequadamente o sinal do resultado e informar também quantas etapas foram necessárias para concluir o processamento. O operador de multiplicação * não pode ser utilizado durante o cálculo

In [72]:
def processador_embarcado_calculo(valor_a, valor_b):
    etapas = 0
    resultado = 0
    while valor_b > 0:
        if valor_b %2 !=0:
            resultado += valor_a
        valor_a = valor_a + valor_a
        valor_b = valor_b//2
        etapas +=1
    print(f"Foram necessárias: {etapas} etapas")
    return resultado

In [79]:
valor_a = int(input("Informe o valor de a: "))
valor_b = int(input("Informe o valor de b: "))
if valor_a < 0 and valor_b < 0:
    valor_a = valor_a * (-1)
    valor_b = valor_b * (-1)
    resultado_calculo = processador_embarcado_calculo(valor_a, valor_b)
    print(f"O resultado do cálculo é: {resultado_calculo}")
elif valor_b < 0:
    valor_b = valor_b * (-1)
    resultado_calculo = processador_embarcado_calculo(valor_a, valor_b)
    resultado_calculo = resultado_calculo * (-1)
    print(f"Como apenas um dois valores são negativos o cálculo é: {resultado_calculo}")
elif valor_a < 0:
    valor_a = valor_a * (-1)
    resultado_calculo = processador_embarcado_calculo(valor_a, valor_b)
    resultado_calculo  = resultado_calculo * (-1)
    print(f"Como apenas um dos dois valores são negativos, o cálculo é negativo: {resultado_calculo}")
else:
    print(f"Resultado do Cálculo: {processador_embarcado_calculo(valor_a, valor_b)}")


Foram necessárias: 2 etapas
Como apenas um dos dois valores são negativos, o cálculo é negativo: -14


### 📑 **Questões Fechadas:** 

#### **1.** Uma empresa de logística utiliza uma esteira automatizada para separar encomendas antes do carregamento dos veículos. Em condições normais, a operação pode utilizar até seis turnos de processamento, mas esse limite é reduzido quando ocorrem situações que indicam baixa eficiência. Uma leitura negativa representa falha do sensor e faz com que um turno previsto seja cancelado. Além disso, depois que mais de 100 encomendas já tiverem sido processadas, uma operação realizada em turno par também provoca a redução do limite de turnos, pois o sistema passa a operar em modo de economia. Independentemente dessas regras, a esteira deve ser desligada assim que 150 encomendas forem processadas.

#### Uma implementação experimental desse controle foi escrita da seguinte forma:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_1_aberta.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### Considerando o comportamento desse programa, avalie as asserções a seguir e a relação proposta entre elas

- I. O processamento pode terminar antes da realização dos seis turnos inicialmente previstos mesmo que a meta de 150 encomendas ainda não tenha sido alcançada.
- **PORQUE**

- II. A variável limite_turnos, que participa diretamente da condição do while, pode ser modificada durante a própria repetição em mais de uma situação prevista pelo algoritmo.

#### A respeito dessas asserções, assinale a alternativa correta.

- A) As asserções I e II são verdadeiras, e II justifica corretamente I.
- B) As asserções I e II são verdadeiras, mas II não justifica I.
- C) A asserção I é verdadeira, e II é falsa.
- D) A asserção I é falsa, e II é verdadeira.
- E) As asserções I e II são falsas.

---

#### Análise da Asserção I:

- *"O processamento pode terminar antes da realização dos seis turnos inicialmente previstos mesmo que a meta de 150 encomendas ainda não tenha sido alcançada."*

#### VERDADEIRA.Depende da quantidade do Lote informado pelo usuário: O loop para quando turno <= limite_turnos é falso. Como limite_turnos pode ser reduzido em duas situações:

- Quando lê um valor negativo (lote < 0)

- Quando já processou mais de 100 e está em turno par (processado > 100 and turno % 2 == 0)

#### Se isso acontecer, limite_turnos diminui e o loop pode terminar antes do turno 6, mesmo sem atingir 150.


#### Exemplo prático:

- Entrada: -1, -1, -1, -1, -1, -1

- Inicia com limite = 6

- A cada negativo, limite diminui em 1

- Após 6 negativos: limite = 0, turno = 7

- Condição turno <= limite_turnos (7 <= 0) é falsa → sai do loop sem processar 150

#### Análise da Asserção II:

- *"A variável limite_turnos, que participa diretamente da condição do while, pode ser modificada durante a própria repetição em mais de uma situação prevista pelo algoritmo."*

#### **VERDADEIRA.** A variável limite_turnos é modificada em DUAS situações diferentes dentro do loop:

- Situação	Linha do código
- 1ª: Leitura de valor negativo	limite_turnos = limite_turnos - 1 (dentro do if lote < 0)
- 2ª: Processado > 100 e turno par	limite_turnos = limite_turnos - 1 (dentro do if processado > 100 and turno % 2 == 0)

#### Ambas as situações reduzem limite_turnos, que é justamente uma das condições de parada do while.

#### Relação entre I e II:
- A II justifica corretamente a I porque:

- A I afirma que o programa pode terminar antes dos 6 turnos mesmo sem atingir a meta

- A II explica o mecanismo que causa isso: limite_turnos pode ser reduzido dentro do loop

#### Se limite_turnos diminui, a condição turno <= limite_turnos pode falhar mais cedo, interrompendo o processamento antes do turno 6

---

#### 2. Um terminal de triagem recebe códigos enviados por diferentes sensores de uma fábrica. Os códigos A e B representam leituras reconhecidas pelo sistema, enquanto a palavra fim encerra a operação. Por segurança, no máximo cinco leituras reconhecidas devem ser processadas em uma mesma execução. Códigos diferentes desses valores deveriam ser apenas registrados como desconhecidos, permitindo que a análise prosseguisse normalmente.

#### Avalie as asserções a seguir e a relação proposta entre elas.

- I. Se o primeiro código informado for X, o programa poderá permanecer indefinidamente no laço de repetição e nunca solicitar uma nova entrada.

#### PORQUE

- II. No fluxo correspondente a um código desconhecido, nenhuma das variáveis que participam da condição de continuidade do while é modificada.

#### A respeito dessas asserções, assinale a alternativa correta.

- A) As asserções I e II são verdadeiras, e II justifica corretamente I.
- B) As asserções I e II são verdadeiras, mas II não justifica I.
- C) A asserção I é verdadeira, e II é falsa.
- D) A asserção I é falsa, e II é verdadeira.
- E) As asserções I e II são falsas.

#### Durante os testes, foi utilizada a seguinte implementação:

---

#### **3.** Uma central de monitoramento recebe medições de consumo energético realizadas ao longo de um dia. Inicialmente, o sistema considera como crítica qualquer leitura maior ou igual a 10 unidades. 
#### Entretanto, esse limite pode ser reduzido sempre que o saldo interno do monitoramento se torna negativo. Uma mesma leitura também pode produzir efeitos diferentes dependendo de sua paridade, pois valores pares provocam uma penalização associada à posição da medição, enquanto valores ímpares produzem um pequeno acréscimo.
#### A equipe responsável pelo sistema implementou o seguinte código:


<div align="center">
<img 
    src="print_abertas_lista_1_02/quetao_3_fechada.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### Avalie as asserções a seguir e a relação proposta entre elas.

#### I. Uma única leitura pode, na mesma iteração, ser contabilizada como acima do limite e também sofrer a atualização correspondente à sua paridade.

#### - *PORQUE*

#### II. As duas primeiras decisões existentes no corpo do for são estruturas condicionais independentes, e não ramos mutuamente exclusivos de uma única decisão.

#### A respeito dessas asserções, assinale a alternativa correta.

- A) As asserções I e II são verdadeiras, e II justifica corretamente I.
- B) As asserções I e II são verdadeiras, mas II não justifica I.
- C) A asserção I é verdadeira, e II é falsa.
- D) A asserção I é falsa, e II é verdadeira.
- E) As asserções I e II são falsas

#### Análise da Asserção I:

- *"Uma única leitura pode, na mesma iteração, ser contabilizada como acima do limite e também sofrer a atualização correspondente à sua paridade."*

#### VERDADEIRA. Sim! Como as duas estruturas condicionais são independentes (não são if/elif/else), uma leitura pode passar pelas duas.

#### Exemplo prático:
- limite = 10

- leitura = 12 (maior que 10 e par)

#### - Passo	O que acontece

In [ ]:
if leitura >= limite:          # Primeiro if
    # bloco 1

if leitura % 2 == 0:           # Segundo if (independente!)
    # bloco 2
else:
    # bloco 3

#### São dois if separados:


- O primeiro verifica se a leitura é maior/igual ao limite

- O segundo verifica a paridade da leitura

#### Eles NÃO estão ligados por elif ou else, portanto:

- Não são mutuamente exclusivos

- Ambos podem ser executados na mesma iteração

- São completamente independentes

#### Relação entre I e II:

- A II justifica corretamente I porque:

- A I afirma que uma mesma leitura pode passar por ambas as condições

- A II explica o mecanismo que permite isso: as duas decisões são independentes

- Se fossem mutuamente exclusivas (um if/elif), isso não seria possível

- Como são independentes, ambas podem ser executadas na mesma iteração

#### Comparativo para clareza:
#### Estrutura atual (independentes):

In [ ]:
if leitura >= limite:
    # executa isso
if leitura % 2 == 0:
    # executa isso também

#### Conclusão:

- ✅ I é verdadeira (leitura pode ser contabilizada nas duas condições)

- ✅ II é verdadeira (as decisões são independentes)

- ✅ II justifica I (explica o mecanismo que permite a ocorrência descrita em I)

#### Alternativa correta: A) As asserções I e II são verdadeiras, e II justifica corretamente I.

---

#### 4. Um sistema de acesso temporário é utilizado em um laboratório que contém equipamentos de alto custo. O usuário recebe um código de autorização e possui, no máximo, quatro oportunidades para informar um código válido. A palavra LIBERAR concede acesso imediatamente, enquanto CANCELAR representa uma desistência voluntária e também encerra o processo. Códigos incorretos aumentam uma penalidade interna: códigos com menos de seis caracteres acrescentam dois pontos e os demais acrescentam um ponto. Ao atingir cinco pontos de penalidade, o sistema deve bloquear novas tentativas, mesmo que o limite de quatro entradas ainda não tenha sido alcançado.

#### O código abaixo implementa esse comportamento, porém a condição do laço foi removida:


<div align="center">
<img 
    src="print_abertas_lista_1_02/quetao_4_fechada.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### Para que o comportamento do programa corresponda ao descrito, a expressão que deve substituir a lacuna é:

- A) codigo != "LIBERAR"and codigo != "CANCELAR"and tentativas < 4 and not bloqueado
- B) codigo != "LIBERAR"or codigo != "CANCELAR"and tentativas < 4 and not bloqueado
- C) codigo != "LIBERAR"and codigo != "CANCELAR"and tentativas <= 4
- D) codigo == "LIBERAR"and tentativas < 4 and not bloqueado
- E) codigo != "LIBERAR"and codigo != "CANCELAR"or not bloqueado

#### Analisando a lógica do problema:

- Condições de parada do loop:

    - Digitar "LIBERAR" → acesso autorizado

    - Digitar "CANCELAR" → operação cancelada

    - Atingir 5 de penalidade → bloqueado = True

    - Exceder 4 tentativas → limite de tentativas

#### O loop deve continuar ENQUANTO TODAS as condições abaixo forem verdadeiras:

- Não digitou "LIBERAR"

- Não digitou "CANCELAR"

- Tentativas < 4 (máximo de 4 oportunidades)

- Não está bloqueado

#### **Analisando cada alternativa:**
  - Alternativa A) ✅ CORRETA

In [ ]:
codigo != "LIBERAR" and codigo != "CANCELAR" and tentativas < 4 and not bloqueado

#### Tradução: Continue enquanto:

- Código NÃO é "LIBERAR" E

- Código NÃO é "CANCELAR" E

- Tentativas < 4 E

- Não está bloqueado

#### Funciona perfeitamente para todos os casos!

#### E por que não poderia ser a Letra -B)? Ordem de precedência dos operadores lógicos no Python e Teste Lógico

In [ ]:
codigo != "LIBERAR" or codigo != "CANCELAR" and tentativas < 4 and not bloqueado

#### **Problema:** Por causa da precedência de operadores em Python (and tem precedência sobre or), a expressão é interpretada como:



In [ ]:
codigo != "LIBERAR" or (codigo != "CANCELAR" and tentativas < 4 and not bloqueado)

#### Teste com codigo = "LIBERAR":

    - codigo != "LIBERAR" = False

    - codigo != "CANCELAR" = True

    - Expressão: False or (True and ...) → True (por causa do or!)

### Resultado: O loop NÃO para quando digita "LIBERAR"!


<div align="center">
<img 
    src="print_abertas_lista_1_02/resposta_4_fechada.png" 
    alt="Gráfico da análise"
    width="900"
>
</div>



---

#### 5. Um laboratório registra sucessivas medições de temperatura obtidas por sensores instalados em uma câmara experimental. Valores negativos representam leiturasinválidas e devem ser desconsiderados. Para as leituras válidas, o sistema calcula a média, contabiliza quantas ficaram acima de 30 graus e registra a maior temperatura observada. Durante um teste em que todos os sensores apresentaram falha, o programa foi encerrado antes de produzir o relatório.

#### Considere a implementação utilizada:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_5_fechada.png" 
    alt="Gráfico da análise"
    width="600"
>
</div>
<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_5_alternativas_1.png" 
    alt="Gráfico da análise"
    width="600"
>
</div>

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_5_alternativas_2.png" 
    alt="Gráfico da análise"
    width="600"
>
</div>

#### Analisando cada alternativa:

#### **Alternativa A)** ✅ CORRETA

- if validas > 0:
    - media = soma / validas
    - print("Media", media)
    - print("Maior", maior)
    - print("Acima de 30", acima)
- else:
    - print("Nenhuma leitura valida")

#### Por que está correta?

- <u>Verifica se existe pelo menos uma leitura válida (validas > 0)</u>

- Se sim, calcula a média normalmente e exibe todos os resultados

- Se não, exibe uma mensagem informativa em vez de gerar erro

- Protege contra divisão por zero sem distorcer os dados.

### **Por que não poderia ser a letra b-)?**

#### Alternativa B) ❌ INCORRETA

- if soma > 0:
    - media = soma / validas
- else:
    - media = 0
- print("Media", media)

#### Problemas:

- Divisão por zero ainda pode ocorrer!

- Se soma = 0 e validas = 0, o else é executado e media = 0, não dá erro

- Mas se soma = 0 e validas = 5 (cinco leituras que somam 0, ex: 1, -1, 2, -2, 0), o if é executado: 0 / 5 = 0 (ok)

### **PORÉM:** Se soma = 5 e validas = 0 (impossível, porque soma só é > 0 se houver válidas)

  - Distorce o resultado!

#### Se todas as leituras forem inválidas (soma = 0, validas = 0), media = 0

### Mas a média não é 0 - não existem dados! Deveria exibir uma mensagem de erro

#### Isso engana o usuário, fazendo parecer que a média foi 0

#### Não resolve o problema de maior e acima

#### Essas variáveis continuariam com valores padrão (0) mesmo sem leituras válidas

#### O relatório seria enganoso

---

#### **6.** Um sistema de certificação numérica recebe identificadores inteiros e utiliza um teste de primalidade antes de autorizar determinados códigos. Para diminuir a quantidade de divisões realizadas, a equipe decidiu testar possíveis divisores somente enquanto o quadrado do divisor atual não ultrapassar o número analisado. Durante a validação, entretanto, códigos como 49 e 121 passaram a ser classificados como primos. O algoritmo implementado foi:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_6_fechada.png" 
    alt="Gráfico da análise"
    width="900"
>
</div>


In [3]:
6%3

0

In [ ]:
n =

#### Para corrigir a classificação dos quadrados perfeitos sem abandonar a estratégia utilizada, a condição do laço deve ser substituída por:

- A) divisor * divisor <= numero and primo
- B) divisor * divisor >= numero and primo
- C) divisor <= numero and not primo
- D) divisor * divisor < numero or primo
- E) divisor != numero and primo


### Início: divisor = 2, primo = True, testes = 0.

- Loop 1: 2*2 = 4 < 49 → entra. Testa 49 % 2 ≠ 0. Incrementa divisor → 3. testes = 1.

- Loop 2: 3*3 = 9 < 49 → entra. Testa 49 % 3 ≠ 0. Incrementa divisor → 4. testes = 2.

- Loop 3: 4*4 = 16 < 49 → entra. Testa 49 % 4 ≠ 0. Incrementa divisor → 5. testes = 3.

- Loop 4: 5*5 = 25 < 49 → entra. Testa 49 % 5 ≠ 0. Incrementa divisor → 6. testes = 4.

- Loop 5: 6*6 = 36 < 49 → entra. Testa 49 % 6 ≠ 0. Incrementa divisor → 7. testes = 5.

- Loop 6: 7*7 = 49 < 49 → condição falsa, sai do loop.

#### Resultado:

#### primo nunca foi alterado → continua True.

#### Impressão: "Numero primo".

####  "Testes 5".

#### --> Nenhum divisor encontrado, então não imprime divisor.

#### Resumindo:


No sexto loop, com divisor * divisor sendo <  numeron , a condição falha porque é < e não <=, oara entrar novamente no loop e modificar a variável booleana e obtendo como resposta um número NÃO Primo E 6 casos de teste

Por isso, 49 é classificado como primo erroneamente.

#### -Resposta: a-): divisor * divisor <= numero and primo

---

#### 7. Uma aplicação experimental analisa combinações de posições de um tabuleiro virtual.

#### Cada linha possui uma quantidade diferente de posições válidas e cada posição gera um valor obtido pelo produto de seus índices. Dependendo desse valor e da soma dos índices, a aplicação realiza atualizações distintas em um acumulador parcial. Paralelamente, uma string é construída para registrar qual regra foi aplicada em cada posição, funcionando como um histórico compacto da execução. Ao terminar cada linha, o sistema mantém apenas o maior resultado parcial já encontrado.

#### Considere a execução do programa para n = 4:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_7_fechada.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### A partir da análise do programa, avalie as afirmações a seguir.

#### **I.** Os valores de parcial produzidos ao final das quatro iterações do laço externo são, respectivamente, −1, −1, 18 e 7.

#### **II.** O valor final de total é 18, embora o último valor calculado para parcial seja 7.

#### **III.** A string armazenada em marcadores possui quatro caracteres ao final da execução, pois o laço externo executa quatro vezes.

#### **IV.** A condição presente no elif somente é avaliada nas posições em que valor % 3 == 0 é falsa.

#### É correto apenas o que se afirma em:

#### Execuções:

#### Iteração 1 → i = 1
  - parcial = 0  
  - Loop interno: j = 1

valor = 1

valor % 3 == 0? → não

(i + j) % 2 == 0? → (1+1=2) sim
→ parcial = 0 - 1 = -1  
→ marcadores = "B"

#### Fim da linha:
  - parcial = -1, total = 0  
  - Print: Linha 1 Parcial -1

#### Iteração 2 → i = 2
  - parcial = 0  
  - Loop interno: j = 1, 2

j = 1 → valor = 2

não múltiplo de 3

(2+1=3) não é par
→ parcial = 0 + 1 = 1  
→ marcadores = "BC"

j = 2 → valor = 4

não múltiplo de 3

(2+2=4) é par
→ parcial = 1 - 2 = -1  
→ marcadores = "BCB"

#### Fim da linha:
  - parcial = -1, total = 0  
  - Print: Linha 2 Parcial -1

####  Iteração 3 → i = 3
  - parcial = 0  
  - Loop interno: j = 1, 2, 3

j = 1 → valor = 3 → múltiplo de 3
→ parcial = 0 + 3 = 3  
→ marcadores = "BCBA"

j = 2 → valor = 6 → múltiplo de 3
→ parcial = 3 + 6 = 9  
→ marcadores = "BCBAA"

j = 3 → valor = 9 → múltiplo de 3
→ parcial = 9 + 9 = 18  
→ marcadores = "BCBAAA"

#### Fim da linha:
  - parcial = 18, total = 18  
  - Print: Linha 3 Parcial 18

#### Iteração 4 → i = 4
parcial = 0  
Loop interno: j = 1, 2, 3, 4

j = 1 → valor = 4 → não múltiplo de 3, (4+1=5) ímpar
→ parcial = 0 + 1 = 1  
→ marcadores = "BCBAAAC"

j = 2 → valor = 8 → não múltiplo de 3, (4+2=6) par
→ parcial = 1 - 2 = -1  
→ marcadores = "BCBAAACB"

j = 3 → valor = 12 → múltiplo de 3
→ parcial = -1 + 12 = 11  
→ marcadores = "BCBAAACBA"

j = 4 → valor = 16 → não múltiplo de 3, (4+4=8) par
→ parcial = 11 - 4 = 7  
→ marcadores = "BCBAAACBAB"

#### Fim da linha:
  - parcial = 7, total = 18  
  - Print: Linha 4 Parcial 7

#### 🔹 Resultado final

  - Código
  - Maior 18
  - Registro BCBAAACBAB


A partir da análise do programa, avalie as afirmações a seguir.

- I. Os valores de parcial produzidos ao final das quatro iterações do laço externo são, respectivamente, −1, −1, 18 e 7. CORRETO!!!

- II. O valor final de total é 18, embora o último valor calculado para parcial seja 7. CORRETO!!!

- III. A string armazenada em marcadores possui quatro caracteres ao final da execução, pois o laço externo executa quatro vezes. ERRADO! TEM  LEN(MARCADORES) == 10

- IV. A condição presente no elif somente é avaliada nas posições em que valor % 3 == 0 é falsa.

#### É correto apenas o que se afirma em:

#### C-) I, II & IV

In [4]:
n = 4

for i in range(1,n+1):
    print("Valor i: ", i)
    for j in range(1, i+1):
        print("Valor j: ", j)

Valor i:  1
Valor j:  1
Valor i:  2
Valor j:  1
Valor j:  2
Valor i:  3
Valor j:  1
Valor j:  2
Valor j:  3
Valor i:  4
Valor j:  1
Valor j:  2
Valor j:  3
Valor j:  4


#### 8. Um sistema interno de conferência utiliza frases de autorização que podem conter espaços para facilitar sua leitura pelos operadores. Antes de efetuar a validação, o sistema gera uma nova string ignorando esses espaços. Em seguida, duas posições são movimentadas simultaneamente a partir das extremidades do texto, comparando pares de caracteres até que seja encontrada uma diferença ou até que não existam mais pares a comparar. A equipe deseja verificar se a implementação evita comparações desnecessárias e se trata corretamente textos de comprimento ímpar. 
#### Considere que a entrada informada seja anotaram a data da maratona e analise o código:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_8_fechada.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>

#### -Primeiro For remove os espaços:

#### **1.** Removendo os espaços

#### Texto original (o que está na imagem é "anotaram a data da maratona", apesar do enunciado mostrar algo ligeiramente diferente — vou usar o que aparece na figura):

#### "anotaramadatadamaratona": len(limpo) == 23

#### Contando os caracteres sem espaços a string "anotaramadatadamaratona", os índices são: 

#### *Lembrar sobre acesso de índices de caractereses funciona semelhante ao acesso de listas, começando do 0*

- Código
- 0 → a
- 1 → n
- 2 → o
- 3 → t
- 4 → a
- 5 → r
- 6 → a
- 7 → m
- 8 → a
- 9 → d
- 10 → a
- 11 → t
- 12 → a
- 13 → d
- 14 → a
- 15 → m
- 16 → a
- 17 → r
- 18 → a
- 19 → t
- 20 → o
- 21 → n
- 22 → a

#### Tamanho da Palavra = 23 len(limpo)

#### Tamanho da variável "fim" = len(limpoo) -1 == 22



#### 2. Verificando se é palíndromo

- String limpa: anotaramadatadmaratona

- Invertendo: anotaramadatadmaratona → mesma coisa? Vamos ver:

#### O comportamento das variáveis inicio e fim no loop se diz respeito ao acréscimo de +1 a início, que começa em 0 e -1 a fim, que começa com 22 a da fim de execução de um Loop
    - Dentro do While,  comparacoes vai sendo acrescida em +1 , e a string limpo, é comparada oela condicional 'if' com cada CARACTERE do começo (início) ao fim da string,como exemplo: limpo(inicio(0==a)  != limpo(fim(22==a))): FALSO!!!! Sendo acrescida de +1 para início e decrementada em -1 para fim


- posição 0 (a) vs posição 22 (a) ✅

- posição 1 (n) vs posição 21 (n) ✅

- posição 2 (o) vs posição 20 (o) ✅

- posição 3 (t) vs posição 19 (t) ✅

- posição 4 (a) vs posição 18 (a) ✅

- posição 5 (r) vs posição 17 (r) ✅

- posição 6 (a) vs posição 16 (a) ✅

- posição 7 (m) vs posição 15 (m) ✅

- posição 8 (a) vs posição 14 (a) ✅

- posição 9 (d) vs posição 13 (d) ✅

- posição 10 (a) vs posição 12 (a) ✅

- posição 11 (t) → centro, não é comparado (inicio = fim = 11, e inicio < fim é falso)

#### Então são 11 comparações (pares de 0-22, 1-21, ..., 10-12). ✅ Afirmação II correta.

#### 3. O programa imprime?

- Como todas as comparações deram certo, valido permanece True, então imprime "Codigo aceito". ✅ Afirmação III correta.

### *4. O caractere central é comparado?*

### O laço while inicio < fim and valido para quando inicio == fim (posição 11). Como a condição usa < (e não <=), o caractere central não é comparado consigo mesmo. ❌ Afirmação IV incorreta.

#### Conclusão

- Corretas: I, II e III → alternativa C.

##### Sua justificativa sobre a IV está certíssima:

#### "IV está errada pois o loop termina antes de comparar o caractere com ele mesmo, pois inicio=11 e fim=11 e inicio < fim é FALSO, para que isso ocorrer deveria ser <=."

#### É exatamente isso. O uso de < em vez de <= evita comparações desnecessárias com o caractere central — o que é justamente o comportamento eficiente que o enunciado menciona ("verificar se a implementação evita comparações desnecessárias").

✅ Resposta: C

---

#### **9.** Um sistema de sincronização precisa determinar o maior intervalo que divide exatamente dois períodos de execução. Para isso, foi implementada uma versão do algoritmo de Euclides. Em cada etapa, o segundo valor é utilizado como divisor do primeiro; o resto encontrado passa a ser o novo segundo valor, enquanto o antigo divisor assume a posição do primeiro. O processo continua até que o resto se torne zero. Além do resultado, a aplicação registra quantas transformações foram necessárias e acumula os restos intermediários para fins de auditoria.

#### Considere as entradas 84 e 30 para o código abaixo:

<div align="center">
<img 
    src="print_abertas_lista_1_02/questao_9_fechada.png" 
    alt="Gráfico da análise"
    width="700"
>
</div>


#### Avalie as afirmações.

- I. O corpo do while é executado três vezes para as entradas fornecidas.
- II. Ao término do algoritmo, a contém 6 e b contém 0.
- III. Substituir a condição b != 0 por a != b preservaria o comportamento do algoritmo para quaisquer dois inteiros positivos.
- IV. Para entradas inteiras positivas, a sequência de restos produzida pelo algoritmo garante que o segundo operando eventualmente alcance zero.

#### É correto apenas o que se afirma em:

- A) I e III.
- B) II e III.
- C) I, II e IV.
- D) I, III e IV.
- E) II, III e IV.

#### Simulação:

- Vamos rodar o código mentalmente para as entradas a = 84 e b = 30:

#### Inicialização:


- a = 84

- b = 30

- passos = 0

- soma_restos = 0

#### Passo 1 (Primeira execução do while):

-  Condição: b != 0 (30 != 0) -> Verdadeiro

- resto = 84 % 30 = 24

- soma_restos = 0 + 24 = 24

- a = b -> a = 30

- b = resto -> b = 24

- passos = 1

#### Imprime: Passo 1 A 30 B 24

- Passo 2 (Segunda execução do while):

- Condição: b != 0 (24 != 0) -> Verdadeiro

- resto = 30 % 24 = 6

- soma_restos = 24 + 6 = 30

- a = b -> a = 24

- b = resto -> b = 6

- passos = 2

#### Imprime: Passo 2 A 24 B 6

#### Passo 3 (Terceira execução do while):

- Condição: b != 0 (6 != 0) -> Verdadeiro

- resto = 24 % 6 = 0

- soma_restos = 30 + 0 = 30

- a = b -> a = 6

- b = resto -> b = 0

- passos = 3

- Imprime: Passo 3 A 6 B 0

#### Passo 4 (Tentativa de execução):

#### Condição: b != 0 (0 != 0) -> Falso

- O loop encerra.

- Saídas Finais:

#### Resultado 6

#### Passos 3

#### Soma 30

##### Análise das Afirmações

- I. O corpo do while é executado três vezes para as entradas fornecidas. Correta. Conforme simulado acima, o loop rodou nos passos 1, 2 e 3.

- II. Ao término do algoritmo, a contém 6 e b contém 0. Correta. No último passo, a recebeu o valor de b (que era 6) e b recebeu o resto (0).

- III. Substituir a condição b != 0 por a != b preservaria o comportamento do algoritmo para quaisquer dois inteiros positivos.

   - ####  Incorreta (Falsa). Como você bem observou, a lógica do algoritmo depende de b ser o divisor e a ser o dividendo. Se mudássemos a condição para a != b, o loop continuaria executando enquanto os números fossem diferentes, mas o problema é que a atualização das variáveis faria com que o algoritmo entrasse em loop infinito ou falhasse em encontrar o MDC corretamente em certos casos.

    - #### Exemplo: Se a=2 e b=4.


    - #### Com b!=0: resto = 2%4 = 2. a=4, b=2. Próximo passo: resto = 4%2 = 0. a=2, b=0. Para.

    - #### Com a!=b: No primeiro passo, 2 != 4 é verdade. resto = 2. a=4, b=2. Agora, 4 != 2 é verdade. resto = 4%2 = 0. a=2, b=0. Agora, 2 != 0 é verdade. resto = 2%0 -> Erro de divisão por zero! O programa quebraria.

#### Portanto, a afirmação III é falsa.

#### IV. Para entradas inteiras positivas, a sequência de restos produzida pelo algoritmo garante que o segundo operando eventualmente alcance zero.

#### Correta. Esta é uma propriedade fundamental do Algoritmo de Euclides. Como o resto é sempre estritamente menor que o divisor (b), a sequência de valores de b é estritamente decrescente (ex: 30, 24, 6, 0). Como são números inteiros positivos, em algum momento chegará a zero.

#### Conclusão

- As afirmações corretas são I, II e IV. Portanto, a alternativa correta é a C.